File: leetcode/median-of-two-sorted-arrays.js
/*
 * @lc app=leetcode.cn id=4 lang=javascript
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
 var findMedianSortedArrays = (nums1, nums2) => {
    let result = 0;
    let total = nums1.length + nums2.length;
    for (let i = 0; i < total; i++) {
        let v1 = nums1.shift();
        let v2 = nums2.shift();
        let v = 0;
        if (v1 != undefined && v2 == undefined) v = v1;
        else if (v2 != undefined && v1 == undefined) v = v2;
        else if (v1 < v2) {
            v = v1;
            nums2.unshift(v2);
        } else {
            v = v2;
            nums1.unshift(v1);
        }
        if (isMiddle(i, total) > 0) break;
        else if (isMiddle(i, total) == 0) result += v;
    }
    return result / (total % 2 == 1 ? 1 : 2);
}
function isMiddle(index, total) {
    if (total % 2 == 1) {
        return index - parseInt(total / 2);
    } else {
        let left = total / 2 - 1;
        if (index < left) return -1;
        if (index > left + 1) return 1;
        return 0;
    }
}
// @lc code=end

File: leetcode/median-of-two-sorted-arrays.cpp
#include "iostream"
#include "vector"
using namespace std;
/*
 * @lc app=leetcode.cn id=4 lang=cpp
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
class Solution
{
public:
  double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
  {
    int total = nums1.size() + nums2.size();
    auto it1 = nums1.begin();
    auto it2 = nums2.begin();
    double result = 0;
    auto next = [&result](vector<int>::iterator& it) -> int
    {
      int res = *it;
      it++;
      return res;
    };
    for (int i = 0; i < total; i++)
    {
      int v;
      if (it1 != nums1.end() && it2 == nums2.end())
        v = next(it1);
      else if (it2 != nums2.end() && it1 == nums1.end())
        v = next(it2);
      else if (*it1 < *it2)
        v = next(it1);
      else
        v = next(it2);

      int com=isMiddle(i, total);
      if (com > 0)
        break;
      if (com == 0)
        result += v;
    }
    return result / (total % 2 == 1 ? 1 : 2);
  }
  int isMiddle(int index, int total)
  {
    if (total % 2 == 1)
      return index - (total / 2) ;
    else
    {
      int lm = total / 2 - 1;
      if (index < lm)
        return -1;
      if (index > lm + 1)
        return 1;
      return 0;
    }
  }
};
// @lc code=end